infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientRouter.java

```diff
@@ -413,7 +413,8 @@ public synchronized void onConnectionUp(String endpoint) {
      */
```

Comment:
When will this callback be called? Which component detects the connection down?

Reply:
The Transport Plugin is the one detecting connections coming up or down.

Reply:
I am just wondering what happen to the snapshot sender/log entry sender  when there is a connection loss and what happen to the fsm state machine.

Reply:
Currently, what happens is that if connection is lost to the remote leader we move to verify remote leader state (if any other connection is still up) or to wait for connections (if no connection is available). But we don't do anything to the snapshot/log entry senders. So good point, this is something we should add. I'm adding that we stop log replication, can you help confirm if that does the correct configuration of the snapshot/log entry sender? or if something else is needed...

Reply:
I saw that the sender catch all exceptions and called cancel.  There are two options:
1. the sender identify the exception type and generate stop event instead of cancel.
2. when the replication state machine detects the current leader connection loss, put a stop event for FSM.

Reply:
We can do in another PR ...

Reply:
I did the option 2.

Reply:
We need to add try and catch for processReads to catch the exceptions.